src/main/xerus/monstercat/api/Playlist.kt

```diff
@@ -0,0 +1,92 @@
+package xerus.monstercat.api
+
+import javafx.beans.property.SimpleBooleanProperty
+import javafx.collections.FXCollections
+import javafx.collections.ObservableList
+import mu.KotlinLogging
+import xerus.ktutil.javafx.properties.SimpleObservable
+import xerus.ktutil.javafx.properties.bindSoft
+import xerus.monstercat.api.response.Track
+import java.util.*
+import kotlin.random.Random
+import kotlin.random.nextInt
+
+object Playlist {
+	val logger = KotlinLogging.logger { }
+	
+	val tracks: ObservableList<Track> = FXCollections.observableArrayList()
+	val history = ArrayDeque<Track>()
+	val currentIndex = SimpleObservable<Int?>(null).apply {
+		bindSoft({
+			tracks.indexOf(Player.activeTrack.value).takeUnless { it == -1 }
+		}, Player.activeTrack, tracks)
+	}
+	
+	val repeat = SimpleBooleanProperty(false)
+	val shuffle = SimpleBooleanProperty(false)
+	
+	operator fun get(index: Int): Track? = tracks.getOrNull(index)
+	
+	/** This property prevents songs that come from the [history] from being added to it again,
+	 * which would effectively create an infinite loop */
+	var lastPolled: Track? = null
+	
+	fun getPrev(): Track? = history.pollLast()?.also {
+		lastPolled = it
+		logger.debug("Polled $it from History")
+	} ?: currentIndex.value?.let { get(it - 1) }
+	
+	fun getNext() = when {
+		shuffle.value -> getNextTrackRandom()
+		repeat.value && (getNextTrack() == null) -> tracks[0]
+		else -> getNextTrack()
+	}
+	
+	fun addNext(track: Track) {
+		tracks.remove(track)
```

Comment:
I don't like that you always remove it before adding to prevent duplicates. Please find a better way. Also, if you add a single track again, it might be intentional. I don't think this automatic removal of duplications is even expected behaviour.

Reply:
With the current implementation we cannot have two of the same track in the Playlist because the currentIndex is calculated with the currently played track, so having duplicates messes with this.

Reply:
why is the currentIndex calculated and not simply tracked?

Reply:
Because you refactored Playlist.kt a while ago. It wasn't me.

Reply:
Didn't change your implementation

Reply:
past me sometimes does weird things. Anyways, subject for future fixing.

Reply:
Sure, we'll find a way eventually.